### PR TITLE
structure containers in categories

### LINKED ISSUE #739: categories for container tab
Currently, all containers are shown in the side bar of the container tab. In my opinion, the sidebar should show the container category instead (e.g. Wine) for two reasons:
1. If you have many containers, showing them in the side bar (even within a scroll pane) is not really usable.
2. With categories, the container tab can be streamlined with the apps and engines tabs which makes the UI more understandable for the user.

Therefore I propose the following solution:
* Use the sidebar to show the container category: All, Wine, ScummVM etc. (To represent this structure, also the folders should be organized in this way, i.e. move `wineprefix` into a new `containers` folder.)
* Show all containers of the selected category in a miniature view in the center (exactly like for apps).
* Show the container information in the right column (see: #733)

### PR DESCRIPTION
**Pull request by plata**
fixes #739 

container categories

### CODE REVIEW

**Comment by madoar** [comment]
I'm not sure if I like the amount of space this needs.
It looks like this may be quite challenging to view on smaller monitors.

File: phoenicis-containers/src/main/java/org/phoenicis/containers/dto/ContainerCategoryDTO.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+/**
+ * Represents a category of container
+ */
+@JsonDeserialize(builder = ContainerCategoryDTO.Builder.class)
+public class ContainerCategoryDTO {
+    private final String name;
+    private final String description;
+    private final List<ContainerDTO> containers;
+    private String icon;
```
Comment (madoar) [inline phoenicis-containers/src/main/java/org/phoenicis/containers/dto/ContainerCategoryDTO.java]: `URI`?

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/mainwindow/containers/ViewContainers.java [hunk 2]
```diff
@@ -18,62 +18,99 @@
...
 
         this.sideBar.setOnApplyFilter(this::applyFilter);
 
-        this.sideBar.bindContainers(this.containers);
+        this.sideBar.bindCategories(this.sortedCategories);
+        Bindings.bindContent(this.availableContainers.getItems(), this.sortedContainers);
 
-        this.initSelectContainerPane();
+        // set the category selection consumers
+        this.sideBar.setOnCategorySelection(category -> {
```
Comment (madoar) [inline phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/mainwindow/containers/ViewContainers.java]: We don't really need the `{` and `}` here.

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/mainwindow/containers/ViewContainers.java [hunk 3]
```diff
@@ -18,62 +18,99 @@
...
+            showAvailableContainers();
+        });
+        this.sideBar.setOnAllCategorySelection(() -> {
```
Comment (madoar) [inline phoenicis-javafx/src/main/java/org/phoenicis/javafx/views/mainwindow/containers/ViewContainers.java]: We don't really need the `{` and `}` here.

File: phoenicis-containers/src/main/java/org/phoenicis/containers/dto/ContainerCategoryDTO.java (hunk 1 above)
Comment (plata) [inline phoenicis-containers/src/main/java/org/phoenicis/containers/dto/ContainerCategoryDTO.java]: Didn't merge master yet.

**Comment by plata** [comment]
I agree that it takes some space. However, there are several reasons why I believe that it still makes sense:

- similar UI as for apps/engines → more consistent user experience
- enables many containers from different engines (imagine you have Wine, ScummVM, ResidualVM with each ~10 apps and all containers shown in the sidebar)
- with #736 the miniature list will consume a lot less space
- there's no need to have 4 Wine tool buttons in a row; with a floating approach this can take more space vertically

File: phoenicis-containers/src/main/java/org/phoenicis/containers/dto/ContainerCategoryDTO.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    public List<ContainerDTO> getContainers() {
+        return containers;
+    }
+
+    public URI getIcon() {
+        return icon;
+    }
+
+    public static Comparator<ContainerDTO> nameComparator() {
+        return (o1, o2) -> o1.getName().compareToIgnoreCase(o2.getName());
```
Comment (madoar) [inline phoenicis-containers/src/main/java/org/phoenicis/containers/dto/ContainerCategoryDTO.java]: We could do this by using `Comparator.comparing(container -> container.getName().toLowerCase())`, right?
I'm not sure which approach is more elegant.

Comment (plata) [inline phoenicis-containers/src/main/java/org/phoenicis/containers/dto/ContainerCategoryDTO.java]: Don't know. I just copy pasted and renamed `EngineCategoryDTO` ;) Didn't think about changing the code.

**Comment by madoar** [comment]
Looks good to me (especially if we change the container panel to better fit inside the details view), but I think it's better to wait for @qparis to see if this fits to his concept.

**Comment by plata** [comment]
Agree to wait for @qparis.

**Comment by qparis** [comment]
Totally agree with the concept. Good job!

**Pull request merged**